distributed/core.py

```diff
@@ -815,40 +815,48 @@ async def _handle_comm(self, comm):
 
     async def handle_stream(self, comm, extra=None):
         extra = extra or {}
-        logger.info("Starting established connection")
+        logger.info("Starting established connection to %s", comm.peer_address)
 
         closed = False
         try:
             while not closed:
-                msgs = await comm.read()
+                try:
+                    msgs = await comm.read()
+                # If another coroutine has closed the comm, stop handling the stream.
+                except CommClosedError:
```

Comment:
As discussed offline with @fjetter, we should use the `CommClosedError` as a stopping and synchronization mechanism between coroutines that are dealing with the comm. If one of them decides to close the comm, the `CommClosedError` allows us to propagate that without race conditions.

Reply:
FWIW It's not ideal but I believe this is how it was intended at some point. We may need to revise this but that would be a bigger change that is feasible right now.